gtsfm/frontend/verifier/loransac.py

```diff
@@ -0,0 +1,194 @@
+        camera_dict2 = {
+            "model": "SIMPLE_PINHOLE",
+            "width": width,
+            "height": height,
+            "params": [focal_length, cx, cy],
+        }
+        result_dict = pycolmap.essential_matrix_estimation(
+            uv_i1, uv_i2, camera_dict1, camera_dict2, max_error_px=self._estimation_threshold_px
+        )
+        return result_dict
+
+    def verify(
+        self,
+        keypoints_i1: Keypoints,
+        keypoints_i2: Keypoints,
+        match_indices: np.ndarray,
+        camera_intrinsics_i1: Cal3Bundler,
+        camera_intrinsics_i2: Cal3Bundler,
+    ) -> Tuple[Optional[Rot3], Optional[Unit3], np.ndarray, float]:
+        """Performs verification of correspondences between two images to recover the relative pose and indices of
+        verified correspondences.
+
+        Args:
+            keypoints_i1: detected features in image #i1.
+            keypoints_i2: detected features in image #i2.
+            match_indices: matches as indices of features from both images, of shape (N3, 2), where N3 <= min(N1, N2).
+            camera_intrinsics_i1: intrinsics for image #i1.
+            camera_intrinsics_i2: intrinsics for image #i2.
+
+        Returns:
+            Estimated rotation i2Ri1, or None if it cannot be estimated.
+            Estimated unit translation i2Ui1, or None if it cannot be estimated.
+            Indices of verified correspondences, of shape (N, 2) with N <= N3. These are subset of match_indices.
+            Inlier ratio of w.r.t. the estimated model, i.e. the #final RANSAC inliers/ #putatives.
+        """
+        if match_indices.shape[0] < self._min_matches:
+            logger.info("[LORANSAC] Not enough correspondences for verification.")
+            return self._failure_result
+
+        uv_i1 = keypoints_i1.coordinates
+        uv_i2 = keypoints_i2.coordinates
+
+        uv_i1_ = uv_i1[match_indices[:, 0]]
+        uv_i2_ = uv_i2[match_indices[:, 1]]
+
+        if self._use_intrinsics_in_verification:
+            result_dict = self.estimate_essential_matrix(uv_i1_, uv_i2_, camera_intrinsics_i1, camera_intrinsics_i2)
+        else:
+            result_dict = pycolmap.fundamental_matrix_estimation(
+                uv_i1_, uv_i2_, max_error_px=self._estimation_threshold_px
+            )
+
+        success = result_dict["success"]
+        if not success:
+            matrix_type = "Essential" if self.use_intrinsics_in_verification else "Fundamental"
+            logger.info(f"[LORANSAC] {matrix_type} matrix estimation unsuccessful.")
+            return self._failure_result
+
+        num_inliers = result_dict["num_inliers"]
+        inlier_ratio_est_model = num_inliers / match_indices.shape[0]
+
+        # no need to extract the relative pose if we have insufficient inliers.
+        if inlier_ratio_est_model < self._min_allowed_inlier_ratio_est_model:
+            i2Ri1 = None
+            i2Ui1 = None
+            v_corr_idxs = np.array([])
```

Comment:
why is this empty? maybe I am missing something, but dont we have any inliers? (maybe less but is it actually zero?)

Reply:
so i guess my thinking here was that if this threshold is not met, then all of the matches are surely garbage. and thus no inliers can be trusted. so, effectively, it's zero

Reply:
Shouldn't the decision of whether to trust the inliers or not be made outside LORANSAC? I think being functional goes beyond just having non-mutable members. The loransac class has a certain function and its scope should be limited to that. One would expect it to return whatever inliers it found. And that is how pycolmap designed their loransac API as well.

Reply:
I think the inliers are good inliers with respect to the model estimated. It's possible the model is garbage, but that is for the caller to handle and handling it here would be beyond loransac functionality.

Reply:
@akshay-krishnan I see what you mean. LORANSAC should just handle the estimation,and a high-level controller should analyze the results with higher-order logic. However, the tricky thing is that if we can't trust the result bc of low inlier ratios, we want to save time and not convert the F matrix to E matrix, for example.

But the verifier API returns an E matrix. So it isn't super clear how to abstract this away at a higher level, since we would need to go back down to the lower level afterwards.

Reply:
@akshay-krishnan gentle reminder about this comment : - )

Reply:
I am not sure what you mean, what goes wrong if we return the correspondences here? I think that makes more sense API-wise. Like R, t was not estimated because we dont have enough inliers, but hey, if you'd like to know what the correspondences were, here you go.

Reply:
Gotcha. I guess I see it as wasted compute to estimate R,t in that case, since we already know that the correspondences will be a waste with very high confidence. Maybe a minor thing, but over 10k or 100k image pairs maybe it would start to add up.

Reply:
Why do we estimate R,t if the inlier ratio is less?

Reply:
It's actually the other way around -- currently I'm not estimating R,t if the inlier ratio is less than the required threshold